### PR TITLE
Epic RPD Update: Multiple Pipe Layers With 1 Click

### PR DESCRIPTION
**Pull request by 13spacemen**
## About The Pull Request
Adds a button to RPD tgui that lets you build multiple pipe layers with the RPD with 1 click
## Why It's Good For The Game
Saves a lot of time when using multiple layers, especially all 5, and is very useful for SM setups and other convoluted atmos contraptions

https://github.com/tgstation/tgstation/assets/46101244/04993fee-7b35-483f-8cb1-a8b58ff1246a

## Changelog
:cl:
qol: You can now build multiple pipe layers with the RPD, with just 1 click!
/:cl:

### CODE REVIEW

**Comment by MTandi** [comment]
If it means that I now need to do 2 clicks instead of 1 just to change the layer, then this is bad UX.

Layer switching is a much more frequent use case than placing multiple objects on the same tile.

**Comment by MTandi** [comment]
> you can scroll up/down when using single layers, that's unchanged

95% of players don't know about this.

**Comment by GuillaumePrata** [comment]
The only reason I can think about for having multiples layers of the same pipe on a tile is to multilayer for extra volume/throughput.

Pushing players to use multilayers for that is a mistake IMO, the layer system should be focused on letting multiple pipelines exist on the same area without conflicts, not quintuple the efficiency of a singular system.

Using 5 layers for volume is common, but it is ugly as sin, if we want to increase the volume of systems like this, we should make a bigger pipe that clogs all layers sprite wise.

It is extra work, but it pays off for not turning atmos systems even deeper into the pipe soup, the barrier of entry for understanding atmos already is too high and it will become even higher if everyone starts 5 layering the SM.

**Comment by Mothblocks** [comment]
> > If it means that I now need to do 2 clicks instead of 1 just to change the layer, then this is bad UX.
> > Layer switching is a much more frequent use case than placing multiple objects on the same tile.
> 
> you can scroll up/down when using single layers, that's unchanged

Not acceptable, the scroll wheel behavior is not great and I would like for it to not exist. I definitely agree that this is a cold case compared to just switching layers. This should be an extra toggle next to the numbers.

File: code/game/objects/items/rcd/RPD.dm [hunk 1]
```diff
@@ -259,6 +263,13 @@ GLOBAL_LIST_INIT(transit_tube_recipes, list(
 	UnregisterSignal(user, COMSIG_MOUSE_SCROLL_ON)
 	return ..()
 
+/obj/item/pipe_dispenser/proc/pipe_layer_bits2num(layerbits)
```
Comment (Mothblocks) [inline code/game/objects/items/rcd/RPD.dm:266]: 1. This proc name is a mouthful. What does it do?
2. Don't abbreviate names, and use snake_case.

File: code/game/objects/items/rcd/RPD.dm [hunk 2]
```diff
@@ -514,46 +526,58 @@ GLOBAL_LIST_INIT(transit_tube_recipes, list(
 	if(mode & BUILD_MODE)
 		switch(category) //if we've gotten this var, the target is valid
 			if(ATMOS_CATEGORY) //Making pipes
```
Comment (Mothblocks) [inline code/game/objects/items/rcd/RPD.dm:528]: Please split this entire thing into its own proc, it's enormous

File: code/game/objects/items/rcd/RPD.dm [hunk 3]
```diff
@@ -202,8 +204,10 @@ GLOBAL_LIST_INIT(transit_tube_recipes, list(
 	var/reprogram_speed = 0.2 SECONDS
 	///Category currently active (Atmos, disposal, transit)
 	var/category = ATMOS_CATEGORY
-	///Piping layer we are going to spawn the atmos device in
-	var/piping_layer = PIPING_LAYER_DEFAULT
+	///All pipe layers we are going to spawn the atmos devices in
+	var/pipe_layers = PIPE_LAYER(3)
+	///Are we only laying a single layer per click
+	var/single_layer = TRUE
```
Comment (Mothblocks) [inline code/game/objects/items/rcd/RPD.dm:210]: This feels like it shouldn't be a variable on the object and should instead be calculated on use.

File: code/game/objects/items/rcd/RPD.dm [hunk 4]
```diff
@@ -514,46 +526,58 @@ GLOBAL_LIST_INIT(transit_tube_recipes, list(
...
-					if(do_after(user, atmos_build_speed, target = attack_target))
-						playsound(get_turf(src), RPD_USE_SOUND, 50, TRUE)
-						var/obj/item/pipe_meter/new_meter = new /obj/item/pipe_meter(get_turf(attack_target))
-						new_meter.setAttachLayer(piping_layer)
-						if(mode & WRENCH_MODE)
-							new_meter.wrench_act(user, src)
-				else
-					if(recipe.all_layers == FALSE && (piping_layer == 1 || piping_layer == 5))
-						balloon_alert(user, "can't build on this layer!")
+				single_layer = !(pipe_layers & (pipe_layers - 1))
```
Comment (Mothblocks) [inline code/game/objects/items/rcd/RPD.dm:529]: I don't like this math. I'd rather we check that it's 1 or a power of 2.

File: code/game/objects/items/rcd/RPD.dm [hunk 5]
```diff
@@ -637,19 +661,24 @@ GLOBAL_LIST_INIT(transit_tube_recipes, list(
 ///Changes the piping layer when the mousewheel is scrolled up or down.
 /obj/item/pipe_dispenser/proc/mouse_wheeled(mob/source_mob, atom/A, delta_x, delta_y, params)
 	SIGNAL_HANDLER
+	single_layer = !(pipe_layers & (pipe_layers - 1))
+	if(!single_layer)
+		to_chat(source_mob, span_notice("Set to single layer first."))
```
Comment (Mothblocks) [inline code/game/objects/items/rcd/RPD.dm:666]: Balloon alert

File: code/game/objects/items/rcd/RPD.dm [hunk 6]
```diff
@@ -637,19 +661,24 @@ GLOBAL_LIST_INIT(transit_tube_recipes, list(
...
-		piping_layer = min(PIPING_LAYER_MAX, piping_layer + 1)
+		pipe_layers = PIPE_LAYER(min(5, pipe_layer_num + 1))
 	else if(delta_y > 0)
-		piping_layer = max(PIPING_LAYER_MIN, piping_layer - 1)
+		pipe_layers = PIPE_LAYER(max(1, pipe_layer_num - 1))
 	else //mice with side-scrolling wheels are apparently a thing and fuck this up
 		return
 	SStgui.update_uis(src)
-	to_chat(source_mob, span_notice("You set the layer to [piping_layer]."))
+	to_chat(source_mob, span_notice("You set the layer to [pipe_layer_bits2num(pipe_layers)]."))
```
Comment (Mothblocks) [inline code/game/objects/items/rcd/RPD.dm:681]: Should change to a balloon alert (this is one of the reasons the scroll behavior is crap right now)

**Review by Mothblocks — CHANGES_REQUESTED** [review]

**Comment by Deadgebert** [comment]
please make this a toggleable, i don't want to lose my right click pipe to change layer nor do i want to click more than once to change the layer i'm working with. I like this addition but it's so incredibly niche that it shouldn't change the functionality of the RPD as a whole

File: code/game/objects/items/rcd/RPD.dm [hunk 7]
```diff
@@ -369,9 +384,21 @@ GLOBAL_LIST_INIT(transit_tube_recipes, list(
...
 			playeffect = FALSE
-		if("piping_layer")
-			piping_layer = text2num(params["piping_layer"])
+		if("pipe_layers")
+			if(multi_layer)
+				var/selected_layers = text2num(params["pipe_layers"])
+				pipe_layers ^= selected_layers
+				if(pipe_layers == 0)
+					pipe_layers = text2num(params["pipe_layers"])
+			else
```
Comment (Mothblocks) [inline code/game/objects/items/rcd/RPD.dm:393]: Please validate that the layer you're receiving is actually valid.

File: code/game/objects/items/rcd/RPD.dm [hunk 8]
```diff
@@ -369,9 +384,21 @@ GLOBAL_LIST_INIT(transit_tube_recipes, list(
...
+				pipe_layers = text2num(params["pipe_layers"])
 			playeffect = FALSE
+		if("toggle_multi_layer")
+			if(multi_layer)
+				pipe_layers = PIPE_LAYER(max(get_active_pipe_layers()))
+			multi_layer = !multi_layer
+		if("set_latest_pipe_layer")
+			latest_pipe_layer = text2num(params["latest_pipe_layer"])
```
Comment (Mothblocks) [inline code/game/objects/items/rcd/RPD.dm:401]: Same

File: code/game/objects/items/rcd/RPD.dm [hunk 9]
```diff
@@ -616,6 +598,76 @@ GLOBAL_LIST_INIT(transit_tube_recipes, list(
...
+	var/list/pipe_layer_numbers = get_active_pipe_layers()
+	for(var/pipe_layer_num in 1 to length(pipe_layer_numbers))
+		var/continued_build = FALSE
+		var/layer_to_build = pipe_layer_numbers[pipe_layer_num]
+		if(!layer_to_build)
+			continue
+		if(!can_make_pipe)
+			return
+		playsound(get_turf(src), 'sound/machines/click.ogg', 50, TRUE)
+		if (recipe.type == /datum/pipe_info/meter)
```
Comment (Mothblocks) [inline code/game/objects/items/rcd/RPD.dm:627]: Early-continue after this

File: code/game/objects/items/rcd/RPD.dm [hunk 10]
```diff
@@ -616,6 +598,76 @@ GLOBAL_LIST_INIT(transit_tube_recipes, list(
...
+	var/can_make_pipe = check_can_make_pipe(atom_to_target)
+	var/list/pipe_layer_numbers = get_active_pipe_layers()
+	for(var/pipe_layer_num in 1 to length(pipe_layer_numbers))
+		var/continued_build = FALSE
+		var/layer_to_build = pipe_layer_numbers[pipe_layer_num]
+		if(!layer_to_build)
+			continue
+		if(!can_make_pipe)
+			return
+		playsound(get_turf(src), 'sound/machines/click.ogg', 50, TRUE)
```
Comment (Mothblocks) [inline code/game/objects/items/rcd/RPD.dm:626]:
```suggestion
		playsound(get_turf(src), 'sound/machines/click.ogg', 50, vary = TRUE)
```

File: code/game/objects/items/rcd/RPD.dm [hunk 11]
```diff
@@ -616,6 +598,76 @@ GLOBAL_LIST_INIT(transit_tube_recipes, list(
...
+					new_meter.wrench_act(user, src)
+				continued_build = TRUE
+		else
+			if(recipe.all_layers == FALSE && (layer_to_build == 1 || layer_to_build == 5))
+				if(multi_layer)
+					continue
+				balloon_alert(user, "can't build on this layer!")
+				return
+			if(continued_build || do_after(user, atmos_build_speed, target = atom_to_target))
+				if(recipe.all_layers == FALSE && (layer_to_build == 1 || layer_to_build == 5))//double check to stop cheaters (and to not waste time waiting for something that can't be placed)
```
Comment (Mothblocks) [inline code/game/objects/items/rcd/RPD.dm:642]:
```suggestion
				if(!recipe.all_layers && (layer_to_build == 1 || layer_to_build == 5))//double check to stop cheaters (and to not waste time waiting for something that can't be placed)
```

File: code/game/objects/items/rcd/RPD.dm [hunk 12]
```diff
@@ -616,6 +598,76 @@ GLOBAL_LIST_INIT(transit_tube_recipes, list(
...
+		if (recipe.type == /datum/pipe_info/meter)
+			if(continued_build || do_after(user, atmos_build_speed, target = atom_to_target))
+				playsound(get_turf(src), RPD_USE_SOUND, 50, TRUE)
+				var/obj/item/pipe_meter/new_meter = new /obj/item/pipe_meter(get_turf(atom_to_target))
+				new_meter.setAttachLayer(layer_to_build)
+				if(mode & WRENCH_MODE)
+					new_meter.wrench_act(user, src)
+				continued_build = TRUE
+		else
+			if(recipe.all_layers == FALSE && (layer_to_build == 1 || layer_to_build == 5))
```
Comment (Mothblocks) [inline code/game/objects/items/rcd/RPD.dm:636]:
```suggestion
			if(!recipe.all_layers && (layer_to_build == 1 || layer_to_build == 5))
```

File: code/game/objects/items/rcd/RPD.dm [hunk 13]
```diff
@@ -616,6 +598,76 @@ GLOBAL_LIST_INIT(transit_tube_recipes, list(
...
+				if(multi_layer)
+					continue
+				balloon_alert(user, "can't build on this layer!")
+				return
+			if(continued_build || do_after(user, atmos_build_speed, target = atom_to_target))
+				if(recipe.all_layers == FALSE && (layer_to_build == 1 || layer_to_build == 5))//double check to stop cheaters (and to not waste time waiting for something that can't be placed)
+					if(multi_layer)
+						continue
+					balloon_alert(user, "can't build on this layer!")
+					return
```
Comment (Mothblocks) [inline code/game/objects/items/rcd/RPD.dm:646]: Should probably make this a proc so we don't have to update two places at once

File: code/game/objects/items/rcd/RPD.dm [hunk 14]
```diff
@@ -637,19 +689,22 @@ GLOBAL_LIST_INIT(transit_tube_recipes, list(
...
-		piping_layer = min(PIPING_LAYER_MAX, piping_layer + 1)
+		pipe_layers = PIPE_LAYER(min(5, (get_active_pipe_layers()[1] + 1)))
 	else if(delta_y > 0)
-		piping_layer = max(PIPING_LAYER_MIN, piping_layer - 1)
+		pipe_layers = PIPE_LAYER(max(1, (get_active_pipe_layers()[1] - 1)))
 	else //mice with side-scrolling wheels are apparently a thing and fuck this up
 		return
 	SStgui.update_uis(src)
-	to_chat(source_mob, span_notice("You set the layer to [piping_layer]."))
+	balloon_alert(source_mob, "set pipe layer [get_active_pipe_layers()[1]]")
```
Comment (Mothblocks) [inline code/game/objects/items/rcd/RPD.dm:707]:
```suggestion
	balloon_alert(source_mob, "set pipe layer to [get_active_pipe_layers()[1]]")
```

**Review by Mothblocks — CHANGES_REQUESTED** [review]

**Comment by LemonInTheDark** [comment]
I feel like an issue in our treatment of atmos is the pattern of advanced users having a problem, and so putting diagnostics/info/tools in just a window.
The extra crap in the analyzer, the weird sucking option for vents, the SM menu, etc.

I worry that we end up causing information overload by doing this. Part of me wants to have this as an upgrade then, so it's opt in extra info rather then coming by default.

File: code/game/objects/items/rcd/RPD.dm [hunk 15]
```diff
@@ -271,7 +282,7 @@ GLOBAL_LIST_INIT(transit_tube_recipes, list(
 		return SECONDARY_ATTACK_CANCEL_ATTACK_CHAIN
 	if(target.pipe_color && target.piping_layer)
 		paint_color = GLOB.pipe_color_name[target.pipe_color]
-		piping_layer = target.piping_layer
+		pipe_layers = target.piping_layer
```
Comment (FlufflesTheDog) [inline code/game/objects/items/rcd/RPD.dm:285]:
```suggestion
		pipe_layers = PIPE_LAYER(target.piping_layer)
```

File: code/game/objects/items/rcd/RPD.dm [hunk 16]
```diff
@@ -414,23 +442,17 @@ GLOBAL_LIST_INIT(transit_tube_recipes, list(
...
 		attack_target = attack_target.wrench_act(user, src)
 		if(!isatom(attack_target)) //can return null, FALSE if do_after() fails see /obj/machinery/atmospherics/wrench_act()
 			return TRUE
 
-	//make sure what we're clicking is valid for the current category
-	var/static/list/make_pipe_whitelist
-	if(!make_pipe_whitelist)
-		make_pipe_whitelist = typecacheof(list(/obj/structure/lattice, /obj/structure/girder, /obj/item/pipe, /obj/structure/window, /obj/structure/grille))
-	if(istype(attack_target, /obj/machinery/atmospherics) && mode & BUILD_MODE)
-		attack_target = get_turf(attack_target)
```
Comment (FlufflesTheDog) [inline code/game/objects/items/rcd/RPD.dm:432]: The code below relies on attack_target being modified, clicking on a pipe with all modes enabled would previously try to build another pipe, now it tries to reconfigure the pipe

**Comment by RikuTheKiller** [comment]
Might be a byond version problem, but I had an issue where I couldn't switch layers when using the plumbing constructor.
None of the buttons did anything and nothing was checked, but I could still select different buildings just fine.
Also, switching layers via scroll wheel still worked just fine.

**Comment by MTandi** [comment]
> > Might be a byond version problem, but I had an issue where I couldn't switch layers when using the plumbing constructor. None of the buttons did anything and nothing was checked, but I could still select different buildings just fine. Also, switching layers via scroll wheel still worked just fine.
> 
> Could you provide some more details? I went in-game on Terry and it seemed to work fine, saw a lot of people playing Atmos Tech too

He's talking about plumbing constructor. They share the layer selection component with RPD I believe.

File: code/game/objects/items/rcd/RPD.dm [hunk 17]
```diff
@@ -609,6 +596,69 @@ GLOBAL_LIST_INIT(transit_tube_recipes, list(
 			else
 				return ..()
 
+/obj/item/pipe_dispenser/proc/check_can_make_pipe(atom/target_of_attack)
+	//make sure what we're clicking is valid for the current category
+	var/static/list/make_pipe_whitelist
+	if(!make_pipe_whitelist)
+		make_pipe_whitelist = typecacheof(list(/obj/structure/lattice, /obj/structure/girder, /obj/item/pipe, /obj/structure/window, /obj/structure/grille))
```
Comment (Ghommie) [inline code/game/objects/items/rcd/RPD.dm:603]:
```suggestion
	var/static/list/make_pipe_whitelist = typecacheof(list(/obj/structure/lattice, /obj/structure/girder, /obj/item/pipe, /obj/structure/window, /obj/structure/grille))
```
This should work just fine, static lists are filled when the server/world is booted and `typecacheof` doesn't really rely on anything created after that (eg. global lists).

**Pull request merged**